[WebAuthn] Implement JSON serialization methods

pascoej opened the pull request:
#### a34c5293eb0921167c306a88a2c7fc38d2b1ee31
<pre>
[WebAuthn] Implement JSON serialization methods
<a href="https://bugs.webkit.org/show_bug.cgi?id=256856">https://bugs.webkit.org/show_bug.cgi?id=256856</a>
<a href="https://rdar.apple.com/problem/109419228">rdar://problem/109419228</a>

Reviewed by Charlie Wolfe.

This patch implements methods to turn different webauthn dictionaries into
more JSON friendly versions. This was added to make ergonomics for developers
easier in <a href="https://w3c.github.io/webauthn/#dom-publickeycredential-tojson.">https://w3c.github.io/webauthn/#dom-publickeycredential-tojson.</a>

BufferSource is tricky here, because it is currently not defined in IDL to the
BufferSource type, but instead IDLUnion&lt;ArrayBuffer, ArrayBufferView&gt;, which we
maintain. To get around this issue I defined a new IDLBufferSource to be used
for situations where we need JSGenerateToJSObject and JSGenerateToNativeObject.
This is the first instance of an IDL type containing a BufferSource needing both
JSGenerateToJSObject and JSGenerateToNativeObject.

I imported the WPT tests that exercise fromJSON, but the WPT toJSON ones require
virtual authenticator support. This is currently only available with WebDriver,
so I instead added WKTR-friendly test cases for toJSON.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-hid.https-expected.txt:
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-hid.https.html:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/webauthn/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/WEB_FEATURES.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/conditional-mediation.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-abort.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-attachment.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-badargs-authnrselection.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-badargs-challenge.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-badargs-rp.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-badargs-user.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-clientdata.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-cross-origin-iframe.https.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-excludecredentials.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-extensions.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-getpublickey.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-hints.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-large-blob-not-supported.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-large-blob-supported.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-minpinlength.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-passing.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-prf.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-pubkeycredparams.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-resident-key.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-timeout.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/credblob-not-supported.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/credblob-supported.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getclientcapabilities.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getclientcapabilities.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-abort.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-allowcredentials.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-attachment.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-badargs-rpid.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-badargs-userverification.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-extensions.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-hints.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-large-blob-not-supported.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-large-blob-supported.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-passing.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-prf.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-rk-passing.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-timeout.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/helpers.js: Added.
(createCredential):
(assertCredential):
(ab2str):
(parseAuthenticatorData):
(TestCase.this.testFunction):
(TestCase):
(TestCase.prototype.toObject):
(TestCase.prototype.toArgs):
(TestCase.prototype.modify.recursiveSetObject):
(TestCase.prototype.modify):
(TestCase.prototype.doIt):
(TestCase.prototype.testPasses):
(TestCase.prototype.testFails):
(TestCase.prototype.runTest):
(TestCase.prototype.testSetup):
(TestCase.prototype.beforeTest):
(TestCase.prototype.testTeardown):
(TestCase.prototype.afterTest):
(TestCase.prototype.validateRet):
(cloneObject):
(extendObject):
(CreateCredentialsTest):
(CreateCredentialsTest.prototype.validateRet):
(GetCredentialsTest):
(GetCredentialsTest.prototype.addCredential):
(GetCredentialsTest.prototype.testSetup):
(GetCredentialsTest.prototype.validateRet):
(GetCredentialsTest.prototype.setIsResidentKeyTest):
(base64urlEncode):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/idlharness-manual.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/idlharness-manual.https.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/idlharness-manual.https.window.js: Added.
(async idlArray):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/idlharness.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/idlharness.https.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/idlharness.https.window.js: Added.
(async idlArray):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-creation-options-from-json.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-creation-options-from-json.https.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-creation-options-from-json.https.window.js: Added.
(test):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-request-options-from-json.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-request-options-from-json.https.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-request-options-from-json.https.window.js: Added.
(test):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-to-json.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-to-json.https.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-to-json.https.window.js: Added.
(assertObjectKeysEq):
(convertObject):
(authenticatorResponseToJson):
(publicKeyCredentialToJson):
(virtualAuthenticatorPromiseTest.async t):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/remote-desktop-client-override.tentative.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/resources/common-inputs.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/resources/utils.js: Added.
(base64urlEncode):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/resources/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/resources/webauthn-subframe.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/securecontext.http-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/securecontext.http.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/securecontext.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/securecontext.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/signal-all-accepted-credentials.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/signal-current-user-details.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/signal-unknown-credential.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/storecredential.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/webauthn-testdriver-basic.https.html: Added.
* LayoutTests/tests-options.json:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.cpp:
(WebCore::AuthenticationExtensionsClientInputs::toCBOR const):
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.h:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.idl:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputsJSON.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl.
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputsJSON.idl: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.idl.
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.cpp:
(WebCore::AuthenticationExtensionsClientOutputs::toJSON const):
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.h:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.idl:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputsJSON.h: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.h.
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputsJSON.idl: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.idl.
* Source/WebCore/Modules/webauthn/AuthenticationResponseJSON.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl.
* Source/WebCore/Modules/webauthn/AuthenticationResponseJSON.idl: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.idl.
* Source/WebCore/Modules/webauthn/AuthenticatorAssertionResponse.cpp:
(WebCore::AuthenticatorAssertionResponse::toJSON):
* Source/WebCore/Modules/webauthn/AuthenticatorAssertionResponse.h:
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
(WebCore::AuthenticatorAttestationResponse::toJSON):
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.h:
* Source/WebCore/Modules/webauthn/AuthenticatorResponse.cpp:
(WebCore::AuthenticatorResponse::data const):
* Source/WebCore/Modules/webauthn/AuthenticatorSelectionCriteria.h: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.idl.
* Source/WebCore/Modules/webauthn/AuthenticatorSelectionCriteria.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl.
* Source/WebCore/Modules/webauthn/CredentialPropertiesOutput.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/CredentialPropertiesOutput.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredential.cpp:
(WebCore::PublicKeyCredential::toJSON):
(WebCore::fromJSON):
(WebCore::PublicKeyCredential::parseCreationOptionsFromJSON):
(WebCore::PublicKeyCredential::parseRequestOptionsFromJSON):
* Source/WebCore/Modules/webauthn/PublicKeyCredential.h:
* Source/WebCore/Modules/webauthn/PublicKeyCredential.idl:
* Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptions.h:
(): Deleted.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptions.idl:
* Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptionsJSON.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptions.h.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptionsJSON.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl:
* Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptorJSON.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptorJSON.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialEntity.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialEntity.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialParameters.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialParameters.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl:
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptionsJSON.h: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.h.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptionsJSON.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRpEntity.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRpEntity.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialUserEntity.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialUserEntity.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialUserEntityJSON.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialUserEntityJSON.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/RegistrationResponseJSON.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl.
* Source/WebCore/Modules/webauthn/RegistrationResponseJSON.idl: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.idl.
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
(fido::convertRpEntityToCBOR):
(fido::convertUserEntityToCBOR):
(fido::convertParametersToCBOR):
(fido::encodeGetAssertionRequestAsCBOR):
* Source/WebCore/Modules/webauthn/fido/U2fCommandConstructor.cpp:
(fido::convertToU2fRegisterCommand):
(fido::convertToU2fCheckOnlySignCommand):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AXRemoteFrame::initializePlatformElementWithRemoteToken):
* Source/WebCore/bindings/IDLTypes.h:
* Source/WebCore/bindings/js/BufferSource.h:
(WebCore::BufferSource::BufferSource):
* Source/WebCore/bindings/js/JSDOMConvertBufferSource.h:
(WebCore::JSConverter&lt;IDLBufferSource&gt;::convert):
(WebCore::toJS):
* Source/WebCore/bindings/js/JSDOMConvertUnion.h:
(WebCore::Converter&lt;IDLBufferSource&gt;::convert):
* Source/WebCore/bindings/js/JSDOMPromiseDeferredForward.h:
* Source/WebCore/bindings/scripts/IDLParser.pm:
(addBuiltinTypedefs):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(publicKeyCredentialRpEntity):
(publicKeyCredentialUserEntity):
(publicKeyCredentialParameters):
(publicKeyCredentialDescriptors):
(authenticatorSelectionCriteria):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::WebCore::collectTransports):
(WebKit::WebCore::getRpId):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(WebKit::LocalAuthenticatorInternal::produceHashSet):
(WebKit::LocalAuthenticator::makeCredential):
(WebKit::LocalAuthenticator::processClientExtensions):
(WebKit::LocalAuthenticator::continueMakeCredentialAfterUserVerification):
(WebKit::LocalAuthenticator::deleteDuplicateCredential const):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForRegistration):
(WebKit::configureRegistrationRequestContext):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::makeCredential):
(WebKit::CtapAuthenticator::continueSilentlyCheckCredentials):
(WebKit::CtapAuthenticator::continueMakeCredentialAfterResponseReceived):
* Source/WebKit/UIProcess/WebAuthentication/fido/U2fAuthenticator.cpp:
(WebKit::U2fAuthenticator::continueRegisterCommandAfterResponseReceived):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Tools/TestWebKitAPI/Tests/WebCore/CtapRequestTest.cpp:
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestParam)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestParamNoUVNoRK)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestParamUVRequiredButNotSupported)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestParamWithPin)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestRKPreferred)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestRKPreferredNotSupported)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestRKDiscouraged)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestWithLargeBlob)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestWithUnsupportedLargeBlob)):
* Tools/TestWebKitAPI/Tests/WebCore/U2fCommandConstructorTest.cpp:
(TestWebKitAPI::constructMakeCredentialRequest):
(TestWebKitAPI::TEST(U2fCommandConstructorTest, TestConvertCtapMakeCredentialToU2fCheckOnlySign)):
(TestWebKitAPI::TEST(U2fCommandConstructorTest, TestConvertCtapMakeCredentialToU2fCheckOnlySignWithInvalidCredentialType)):
(TestWebKitAPI::TEST(U2fCommandConstructorTest, TestU2fRegisterCredentialAlgorithmRequirement)):
(TestWebKitAPI::TEST(U2fCommandConstructorTest, TestU2fRegisterUserVerificationRequirement)):
(TestWebKitAPI::TEST(U2fCommandConstructorTest, TestU2fRegisterResidentKeyRequirement)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKWebAuthenticationPanel.mm:
(TestWebKitAPI::TEST(WebAuthenticationPanel, PublicKeyCredentialCreationOptionsMinimum)):
(TestWebKitAPI::TEST(WebAuthenticationPanel, PublicKeyCredentialCreationOptionsMaximumDefault)):
(TestWebKitAPI::TEST(WebAuthenticationPanel, PublicKeyCredentialCreationOptionsMaximum1)):
(TestWebKitAPI::TEST(WebAuthenticationPanel, PublicKeyCredentialCreationOptionsMaximum2)):

Canonical link: <a href="https://commits.webkit.org/284627@main">https://commits.webkit.org/284627@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1fda828cd155523eec4c5b882047edf002275da7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70027 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49428 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22780 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74112 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21185 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72144 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57228 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21036 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55577 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14050 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73093 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45033 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60403 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36043 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41693 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17821 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19562 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63617 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18171 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75830 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14252 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17415 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63280 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14288 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60467 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63203 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15535 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11217 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4828 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45231 "Built successfully") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46305 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47576 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46046 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
```diff
@@ -183,6 +184,22 @@ RefPtr<ArrayBuffer> AuthenticatorAttestationResponse::getPublicKey() const
+        base64EncodeToString(clientDataJSON()->span()),
+        base64EncodeToString(getAuthenticatorData()->span()),
+        transports,
+        base64EncodeToString(getPublicKey()->span()),
```
These pointers are nullable, do they need a null check?

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredential.h:
```diff
@@ -27,8 +27,16 @@
+#include "PublicKeyCredentialCreationOptionsJSON.h"
+#include "PublicKeyCredentialRequestOptions.h"
+#include "PublicKeyCredentialRequestOptionsJSON.h"
+#include "RegistrationResponseJSON.h"
```
Some of these look like they can be forward declared

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptionsJSON.h:
```diff
@@ -0,0 +1,63 @@
+#include "PublicKeyCredentialUserEntityJSON.h"
+#include "ResidentKeyRequirement.h"
+#include "UserVerificationRequirement.h"
+#include <wtf/Forward.h>
```
Some of these look like they can be forward declared

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptorJSON.h:
```diff
@@ -0,0 +1,41 @@
+#include <wtf/Forward.h>
+
+namespace WebCore {
+struct PublicKeyCredentialDescriptorJSON {
```
nit - usually there are spaces between the namespace and object declaration

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptorJSON.h:
```diff
@@ -0,0 +1,41 @@
+#if ENABLE(WEB_AUTHN)
+
+#include "IDLTypes.h"
+#include <wtf/Forward.h>
```
I don't think these includes are needed.

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialEntity.h:
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2017 Apple Inc. All rights reserved.
```
2024

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialEntity.h:
```diff
@@ -0,0 +1,41 @@
+
+#include "BasicCredential.h"
+#include "IDLTypes.h"
+#include <wtf/Forward.h>
```
I don't think these includes are needed.

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialEntity.h:
```diff
@@ -0,0 +1,41 @@
+#include <wtf/Forward.h>
+
+namespace WebCore {
+struct PublicKeyCredentialEntity {
```
nit - usually there are spaces between the namespace and object declaration

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialParameters.h:
```diff
@@ -0,0 +1,43 @@
+
+#include "IDLTypes.h"
+#include "PublicKeyCredentialType.h"
+#include <wtf/Forward.h>
```
Only PublicKeyCredentialType.h is needed?

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialParameters.h:
```diff
@@ -0,0 +1,43 @@
+
+struct PublicKeyCredentialParameters {
+    PublicKeyCredentialType type;
+    int64_t alg;
```
Should this be initialized?

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptionsJSON.h:
```diff
@@ -0,0 +1,58 @@
+#include "PublicKeyCredentialUserEntityJSON.h"
+#include "ResidentKeyRequirement.h"
+#include "UserVerificationRequirement.h"
+#include <wtf/Forward.h>
```
You should remove some of these includes

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialRpEntity.h:
```diff
@@ -0,0 +1,40 @@
+
+#include "IDLTypes.h"
+#include "PublicKeyCredentialEntity.h"
+#include <wtf/Forward.h>
```
unneeded includes

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialRpEntity.h:
```diff
@@ -0,0 +1,40 @@
+#include <wtf/Forward.h>
+
+namespace WebCore {
+struct PublicKeyCredentialRpEntity : public PublicKeyCredentialEntity {
```
spaces between namespace and object declaration

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialUserEntityJSON.idl:
```diff
@@ -0,0 +1,34 @@
+    Conditional=WEB_AUTHN,
+] dictionary PublicKeyCredentialUserEntityJSON {
+    required Base64URLString id;
+    required DOMString              name;
```
```suggestion
    required DOMString name;
```

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialUserEntity.h:
```diff
@@ -0,0 +1,42 @@
+#include "BufferSource.h"
+#include "IDLTypes.h"
+#include "PublicKeyCredentialEntity.h"
+#include <wtf/Forward.h>
```
same comment about includes

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialUserEntity.h:
```diff
@@ -0,0 +1,42 @@
+#include <wtf/Forward.h>
+
+namespace WebCore {
+struct PublicKeyCredentialUserEntity : public PublicKeyCredentialEntity {
```
same comment about spacing

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialUserEntityJSON.h:
```diff
@@ -0,0 +1,41 @@
+#if ENABLE(WEB_AUTHN)
+
+#include "IDLTypes.h"
+#include <wtf/Forward.h>
```
same comment about includes

charliewolfe commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialUserEntityJSON.h:
```diff
@@ -0,0 +1,41 @@
+#include <wtf/Forward.h>
+
+namespace WebCore {
+struct PublicKeyCredentialUserEntityJSON {
```
same comment about spacing

charliewolfe commented on Source/WebCore/Modules/webauthn/RegistrationResponseJSON.h:
```diff
@@ -0,0 +1,54 @@
+        String authenticatorData;
+        Vector<String> transports;
+        String publicKey;
+        long long publicKeyAlgorithm;
```
initialize this?

charliewolfe commented on Source/WebCore/bindings/js/JSDOMConvertBufferSource.h:
```diff
@@ -87,13 +88,64 @@ inline JSC::JSValue toJS(JSC::JSGlobalObject* lexicalGlobalObject, JSC::JSGlobal
     return view.wrap(lexicalGlobalObject, globalObject);
 }
 
+template<> struct JSConverter<IDLBufferSource> {
```
Can you really not use any of the existing templated JSConverters for this?

charliewolfe reviewed (approved):
WebAuthn changes lgtm. You may want someone else to look at the bindings changes.

pascoej committed:
[WebAuthn] Implement JSON serialization methods
https://bugs.webkit.org/show_bug.cgi?id=256856
rdar://problem/109419228

Reviewed by Charlie Wolfe.

This patch implements methods to turn different webauthn dictionaries into
more JSON friendly versions. This was added to make ergonomics for developers
easier in https://w3c.github.io/webauthn/#dom-publickeycredential-tojson.

BufferSource is tricky here, because it is currently not defined in IDL to the
BufferSource type, but instead IDLUnion<ArrayBuffer, ArrayBufferView>, which we
maintain. To get around this issue I defined a new IDLBufferSource to be used
for situations where we need JSGenerateToJSObject and JSGenerateToNativeObject.
This is the first instance of an IDL type containing a BufferSource needing both
JSGenerateToJSObject and JSGenerateToNativeObject.

I imported the WPT tests that exercise fromJSON, but the WPT toJSON ones require
virtual authenticator support. This is currently only available with WebDriver,
so I instead added WKTR-friendly test cases for toJSON.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-hid.https-expected.txt:
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-hid.https.html:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/webauthn/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/WEB_FEATURES.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/conditional-mediation.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-abort.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-attachment.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-badargs-authnrselection.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-badargs-challenge.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-badargs-rp.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-badargs-user.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-clientdata.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-cross-origin-iframe.https.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-excludecredentials.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-extensions.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-getpublickey.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-hints.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-large-blob-not-supported.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-large-blob-supported.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-minpinlength.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-passing.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-prf.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-pubkeycredparams.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-resident-key.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/createcredential-timeout.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/credblob-not-supported.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/credblob-supported.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getclientcapabilities.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getclientcapabilities.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-abort.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-allowcredentials.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-attachment.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-badargs-rpid.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-badargs-userverification.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-extensions.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-hints.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-large-blob-not-supported.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-large-blob-supported.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-passing.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-prf.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-rk-passing.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/getcredential-timeout.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/helpers.js: Added.
(createCredential):
(assertCredential):
(ab2str):
(parseAuthenticatorData):
(TestCase.this.testFunction):
(TestCase):
(TestCase.prototype.toObject):
(TestCase.prototype.toArgs):
(TestCase.prototype.modify.recursiveSetObject):
(TestCase.prototype.modify):
(TestCase.prototype.doIt):
(TestCase.prototype.testPasses):
(TestCase.prototype.testFails):
(TestCase.prototype.runTest):
(TestCase.prototype.testSetup):
(TestCase.prototype.beforeTest):
(TestCase.prototype.testTeardown):
(TestCase.prototype.afterTest):
(TestCase.prototype.validateRet):
(cloneObject):
(extendObject):
(CreateCredentialsTest):
(CreateCredentialsTest.prototype.validateRet):
(GetCredentialsTest):
(GetCredentialsTest.prototype.addCredential):
(GetCredentialsTest.prototype.testSetup):
(GetCredentialsTest.prototype.validateRet):
(GetCredentialsTest.prototype.setIsResidentKeyTest):
(base64urlEncode):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/idlharness-manual.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/idlharness-manual.https.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/idlharness-manual.https.window.js: Added.
(async idlArray):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/idlharness.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/idlharness.https.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/idlharness.https.window.js: Added.
(async idlArray):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-creation-options-from-json.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-creation-options-from-json.https.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-creation-options-from-json.https.window.js: Added.
(test):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-request-options-from-json.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-request-options-from-json.https.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-request-options-from-json.https.window.js: Added.
(test):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-to-json.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-to-json.https.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/public-key-credential-to-json.https.window.js: Added.
(assertObjectKeysEq):
(convertObject):
(authenticatorResponseToJson):
(publicKeyCredentialToJson):
(virtualAuthenticatorPromiseTest.async t):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/remote-desktop-client-override.tentative.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/resources/common-inputs.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/resources/utils.js: Added.
(base64urlEncode):
* LayoutTests/imported/w3c/web-platform-tests/webauthn/resources/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/resources/webauthn-subframe.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/securecontext.http-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/securecontext.http.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/securecontext.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/securecontext.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/signal-all-accepted-credentials.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/signal-current-user-details.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/signal-unknown-credential.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/storecredential.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/webauthn/webauthn-testdriver-basic.https.html: Added.
* LayoutTests/tests-options.json:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.cpp:
(WebCore::AuthenticationExtensionsClientInputs::toCBOR const):
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.h:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.idl:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputsJSON.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl.
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputsJSON.idl: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.idl.
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.cpp:
(WebCore::AuthenticationExtensionsClientOutputs::toJSON const):
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.h:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.idl:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputsJSON.h: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.h.
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputsJSON.idl: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.idl.
* Source/WebCore/Modules/webauthn/AuthenticationResponseJSON.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl.
* Source/WebCore/Modules/webauthn/AuthenticationResponseJSON.idl: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.idl.
* Source/WebCore/Modules/webauthn/AuthenticatorAssertionResponse.cpp:
(WebCore::AuthenticatorAssertionResponse::toJSON):
* Source/WebCore/Modules/webauthn/AuthenticatorAssertionResponse.h:
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
(WebCore::AuthenticatorAttestationResponse::toJSON):
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.h:
* Source/WebCore/Modules/webauthn/AuthenticatorResponse.cpp:
(WebCore::AuthenticatorResponse::data const):
* Source/WebCore/Modules/webauthn/AuthenticatorSelectionCriteria.h: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.idl.
* Source/WebCore/Modules/webauthn/AuthenticatorSelectionCriteria.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl.
* Source/WebCore/Modules/webauthn/CredentialPropertiesOutput.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/CredentialPropertiesOutput.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredential.cpp:
(WebCore::PublicKeyCredential::toJSON):
(WebCore::fromJSON):
(WebCore::PublicKeyCredential::parseCreationOptionsFromJSON):
(WebCore::PublicKeyCredential::parseRequestOptionsFromJSON):
* Source/WebCore/Modules/webauthn/PublicKeyCredential.h:
* Source/WebCore/Modules/webauthn/PublicKeyCredential.idl:
* Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptions.h:
(): Deleted.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptions.idl:
* Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptionsJSON.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptions.h.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptionsJSON.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl:
* Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptorJSON.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptorJSON.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialEntity.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialEntity.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialParameters.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialParameters.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl:
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptionsJSON.h: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.h.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptionsJSON.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRpEntity.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRpEntity.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialUserEntity.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialUserEntity.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialUserEntityJSON.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredentialUserEntityJSON.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialDescriptor.idl.
* Source/WebCore/Modules/webauthn/RegistrationResponseJSON.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl.
* Source/WebCore/Modules/webauthn/RegistrationResponseJSON.idl: Copied from Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.idl.
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
(fido::convertRpEntityToCBOR):
(fido::convertUserEntityToCBOR):
(fido::convertParametersToCBOR):
(fido::encodeGetAssertionRequestAsCBOR):
* Source/WebCore/Modules/webauthn/fido/U2fCommandConstructor.cpp:
(fido::convertToU2fRegisterCommand):
(fido::convertToU2fCheckOnlySignCommand):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AXRemoteFrame::initializePlatformElementWithRemoteToken):
* Source/WebCore/bindings/IDLTypes.h:
* Source/WebCore/bindings/js/BufferSource.h:
(WebCore::BufferSource::BufferSource):
* Source/WebCore/bindings/js/JSDOMConvertBufferSource.h:
(WebCore::JSConverter<IDLBufferSource>::convert):
(WebCore::toJS):
* Source/WebCore/bindings/js/JSDOMConvertUnion.h:
(WebCore::Converter<IDLBufferSource>::convert):
* Source/WebCore/bindings/js/JSDOMPromiseDeferredForward.h:
* Source/WebCore/bindings/scripts/IDLParser.pm:
(addBuiltinTypedefs):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(publicKeyCredentialRpEntity):
(publicKeyCredentialUserEntity):
(publicKeyCredentialParameters):
(publicKeyCredentialDescriptors):
(authenticatorSelectionCriteria):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::WebCore::collectTransports):
(WebKit::WebCore::getRpId):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(WebKit::LocalAuthenticatorInternal::produceHashSet):
(WebKit::LocalAuthenticator::makeCredential):
(WebKit::LocalAuthenticator::processClientExtensions):
(WebKit::LocalAuthenticator::continueMakeCredentialAfterUserVerification):
(WebKit::LocalAuthenticator::deleteDuplicateCredential const):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForRegistration):
(WebKit::configureRegistrationRequestContext):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::makeCredential):
(WebKit::CtapAuthenticator::continueSilentlyCheckCredentials):
(WebKit::CtapAuthenticator::continueMakeCredentialAfterResponseReceived):
* Source/WebKit/UIProcess/WebAuthentication/fido/U2fAuthenticator.cpp:
(WebKit::U2fAuthenticator::continueRegisterCommandAfterResponseReceived):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Tools/TestWebKitAPI/Tests/WebCore/CtapRequestTest.cpp:
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestParam)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestParamNoUVNoRK)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestParamUVRequiredButNotSupported)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestParamWithPin)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestRKPreferred)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestRKPreferredNotSupported)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestRKDiscouraged)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestWithLargeBlob)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestWithUnsupportedLargeBlob)):
* Tools/TestWebKitAPI/Tests/WebCore/U2fCommandConstructorTest.cpp:
(TestWebKitAPI::constructMakeCredentialRequest):
(TestWebKitAPI::TEST(U2fCommandConstructorTest, TestConvertCtapMakeCredentialToU2fCheckOnlySign)):
(TestWebKitAPI::TEST(U2fCommandConstructorTest, TestConvertCtapMakeCredentialToU2fCheckOnlySignWithInvalidCredentialType)):
(TestWebKitAPI::TEST(U2fCommandConstructorTest, TestU2fRegisterCredentialAlgorithmRequirement)):
(TestWebKitAPI::TEST(U2fCommandConstructorTest, TestU2fRegisterUserVerificationRequirement)):
(TestWebKitAPI::TEST(U2fCommandConstructorTest, TestU2fRegisterResidentKeyRequirement)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKWebAuthenticationPanel.mm:
(TestWebKitAPI::TEST(WebAuthenticationPanel, PublicKeyCredentialCreationOptionsMinimum)):
(TestWebKitAPI::TEST(WebAuthenticationPanel, PublicKeyCredentialCreationOptionsMaximumDefault)):
(TestWebKitAPI::TEST(WebAuthenticationPanel, PublicKeyCredentialCreationOptionsMaximum1)):
(TestWebKitAPI::TEST(WebAuthenticationPanel, PublicKeyCredentialCreationOptionsMaximum2)):

Canonical link: https://commits.webkit.org/284627@main
